File: data-structures/stack/stack_test.go
package stack

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestStack(t *testing.T) {
	stack := NewStack[int](5)
	assert.Equal(t, uint(5), stack.Capacity())

	firstOk := stack.Push(10)
	assert.True(t, firstOk)
	assert.Equal(t, uint(1), stack.Size())

	stack.Push(20)
	assert.Equal(t, uint(2), stack.Size())

	stack.Push(30)
	stack.Push(40)
	fifthOk := stack.Push(50)
	assert.True(t, fifthOk)
	assert.Equal(t, uint(5), stack.Size())

	sixthOk := stack.Push(60)
	assert.False(t, sixthOk)
	assert.Equal(t, uint(5), stack.Size())

	peeked := stack.Peek()
	assert.Equal(t, 50, *peeked)
	assert.Equal(t, uint(5), stack.Size())

	popped := stack.Pop()
	assert.Equal(t, 50, *popped)
	assert.Equal(t, uint(4), stack.Size())

	// remove all.
	for range 5 {
		stack.Pop()
	}
	assert.True(t, stack.Size() == 0)
}

File: app/data_structures/stack/stack_test.py
from unittest import TestCase
from .stack import Stack


class TestStack(TestCase):
    def test_normal_usage(self) -> None:
        s = Stack[int](capacity=5)
        self.assertEqual(s.capacity, 5)
        for i in range(4):
            s.push(i + 1)

        self.assertEqual(len(s), 4)
        self.assertEqual(s.capacity, 5)

        s.push(5)
        self.assertEqual(5, len(s))
        s.push(6)
        self.assertEqual(6, len(s))
        self.assertEqual(10, s.capacity)

        self.assertEqual(s.peek(), 6)
        self.assertEqual(len(s), 6)
        self.assertEqual(s.pop(), 6)
        self.assertEqual(len(s), 5)

    def test_empty_stack(self) -> None:
        s = Stack[int](10)
        self.assertEqual(len(s), 0)
        self.assertEqual(s.capacity, 10)

        values = list(s.__iter__())
        self.assertEqual(values, [])
        self.assertIsNone(s.pop())
        self.assertIsNone(s.peek())
